Node Conformance Test: Start e2e services in a separate process

Random-Liu opened the pull request:
For https://github.com/kubernetes/kubernetes/issues/30122.

This is the first step of https://github.com/kubernetes/kubernetes/issues/30174.

This PR added a `start-services-only` mode in the test suite, and start all e2e services in a separate process by run the test binary again in `start-services-only` mode.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30116)

<!-- Reviewable:end -->

Random-Liu committed:
Add run-services-mode option, and start e2e services in a separate
process.

vishh commented on test/e2e_node/e2e_node_suite_test.go:
```diff
@@ -47,38 +46,25 @@ import (
 
 func TestE2eNode(t *testing.T) {
 	flag.Parse()
+	if *startServicesOnly {
```
Why is this mode needed?

vishh commented on test/e2e_node/e2e_service.go:
```diff
@@ -33,22 +34,181 @@ import (
+	cmd *exec.Cmd
+}
+
+func NewE2eServices() *E2EServices {
```
Why a public method?

Random-Liu replied:
I planed to move this into a separate `service` package, so make it public to make coming refactor easier. :)

vishh commented on test/e2e_node/e2e_node_suite_test.go:
```diff
@@ -124,45 +118,47 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 
-	framework.TestContext.NodeName = shared.NodeName
+	// Initialize node name here to make sure every test node will get right node name.
+	if framework.TestContext.NodeName == "" {
```
Why is the node name being initialized twice now?

Random-Liu replied:
There are 2 functions in `SynchronizedBeforeSuite`. The first one is only called once, the second one is called on every parallel test node.

We need to initialize node name to start kubelet in the first function now.
However, only the variable initialized in the second function can be accessed by all test nodes, so we have to initialize it again in the second function.

After moving the kubelet start logic out, we can only keep the second initalization.

vishh commented on test/e2e_node/e2e_service.go:
```diff
@@ -33,22 +34,181 @@ import (
+// StartE2eServices start the e2e services in another process, it returns
+// when all e2e services are ready.
+// We want to statically link e2e services into the test binary, but we don't
+// want their glog to pollute the test result. So we run the binary in start-
```
I don't get this. Why can't the test binary start processes in a separate process and then run the tests itself?

vishh commented on test/e2e_node/e2e_service.go:
```diff
@@ -33,22 +34,181 @@ import (
+		return manifestPath, fmt.Errorf("can't get current binary: %v", err)
+	}
+	cmd := exec.Command(testBin,
+		"--start-services-only",
```
I think I get what you are trying to do. Instead how about this:

1. Invoke `e2e-node.test`
1. `e2e-node.test` will internally exec `e2e-node.test apiserver` & `e2e-node.test etcd`
1. `e2e-node.test` will run the tests only the exec'ed processes are running.
1. `e2e-node.test` will evaluate os.Argv[1] to determine if it has to start etcd or apiserver, or run in test mode.

Also consider using env variables instead of flags if you'd like to hide some of the implementation details.

Random-Liu replied:
@vishh Thanks for reviewing!

Using env is a good idea to hide the implementation details! Thanks!! :)

What is the difference between "starting apiserver, etcd and maybe namespace controller in different processes with different command" and "starting one process, and letting it starts all dependencies"?
The later one seems easier to me. The former one can help us redirect logs of different services, but do we care their log? :)

Random-Liu commented:
@vishh I've updated the PR to mark the `start-services-only` flag as "hidden", which will avoid user seeing it.

For the ux problem:
* Firstly, I don't think we should expose "subcommand" to the user. For node conformance test, user should not care about starting apiserver, etcd or something at all, they should only run and test their node. All details should be hide behind.
* Secondly, even for the current test binary, there are a lot of cadvisor flags we can't eliminate. https://github.com/google/cadvisor/issues/1417
It seems that the binary is not ready to be the user facing interface yet, we'd better move on to containerize the test and make it easy to run the container which is our main goal for 1.4.

WDYT?

Random-Liu commented:
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/30399.

Random-Liu commented:
Also add @ronnielai as the reviewer. Thanks a lot for paying time for this! ! :0

Random-Liu commented:
Also add @timstclair  as the reviewer. Thanks a lot! :)

ronnielai commented on test/e2e/framework/test_context.go:
```diff
@@ -69,10 +69,15 @@ type TestContextType struct {
 	DumpLogsOnFailure bool
 	// Name of the node to run tests on (node e2e suite only).
 	NodeName string
+	// TODO(random-liu): Remove following flags.
```
Could you put in the comment about when DisableKubenet will be removed?

ronnielai commented on test/e2e/framework/test_context.go:
```diff
@@ -151,6 +156,10 @@ func RegisterClusterFlags() {
 func RegisterNodeFlags() {
 	flag.StringVar(&TestContext.NodeName, "node-name", "", "Name of the node to run tests on (node e2e suite only).")
+	// TODO(random-liu): Remove kubelet related flags when we move the kubelet start logic out of the test.
+	// TODO(random-liu): Add an flag to pass in all kubelet flags set outside, and automatic config and filter test based on the flags.
```
Could you elaborate on "set outside"? Outside of what?

ronnielai commented on test/e2e_node/e2e_node_suite_test.go:
```diff
@@ -111,6 +99,14 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 	}
 
+	// Initialize node name here, so that the following code can get right node name.
+	if framework.TestContext.NodeName == "" {
```
Could you make it a method since it's called twice?

timstclair commented:
Reviewed 3 of 6 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[test/e2e/framework/test_context.go, line 161 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-iTDaJNBFpc4fXdFB:-KP-iTDbr8xKYe7trHhA:b2xezcc) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e/framework/test_context.go#L161)):*
> ```Go
> 	// TODO(random-liu): Remove kubelet related flags when we move the kubelet start logic out of the test.
> 	// TODO(random-liu): Add an flag to pass in all kubelet flags set outside, and automatic config and filter test based on the flags.
> 	flag.BoolVar(&TestContext.DisableKubenet, "disable-kubenet", false, "If true, start kubelet without kubenet. (default false)")
> ```

Isn't `(default false)` added automatically? If you use `--help` is it shown twice?

---

*[test/e2e_node/e2e_node_suite_test.go, line 53 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-kEFz8U-Ik7NB6Tof:-KP-kEFz8U-Ik7NB6Tog:b-qudfd1) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L53)):*
> ```Go
> 
> var prePullImages = flag.Bool("prepull-images", true, "If true, prepull images so image pull failures do not cause test failures.")
> var startServicesOnly = flag.Bool("start-services-only", false, "If true, only start services (etcd, apiserver), and not run test. (default false)")
> ```

The relationship of `start-services` and `start-services-only` is a little confusing. Would it make more sense to have a `run tests` flag that could be used with `start-services`?

---

*[test/e2e_node/e2e_node_suite_test.go, line 59 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-j3LSF6mR67R7FdKe:-KP-j3LTPVEI1AKG2LQG:b-eqxo03) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L59)):*
> ```Go
> 	framework.RegisterNodeFlags()
> 	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
> 	// Mark the start-services-only flag as hidden to avoid user from using it.
> ```

nit: (grammar) s/avoid/prevent/

---

*[test/e2e_node/e2e_node_suite_test.go, line 61 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-jTlwxx_EJAqaiUKq:-KP-jTlwxx_EJAqaiUKr:b-gv274x) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L61)):*
> ```Go
> 	// Mark the start-services-only flag as hidden to avoid user from using it.
> 	pflag.CommandLine.MarkHidden("start-services-only")
> 	pflag.Parse()
> ```

Flags should not be parsed in init. The problem is, flags could be added at package init time in other packages, and the ordering is hard to track.

---

*[test/e2e_node/e2e_node_suite_test.go, line 124 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-lFjbX35QKT936Rf8:-KP-lFjc6mWFWcsHNVuT:bkcz6lp) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L124)):*
> ```Go
> 
> 	var data []byte
> 	if *startServices {
> ```

Can you call RunE2EServices here, or does this do something different? It's confusing...

---

*[test/e2e_node/e2e_service.go, line 58 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_yA-r1-58:-KP-ljPDmFb-cWMMTE5t:bgat1l5) ([raw file](https://github.com/kubernetes/kubernetes/blob/bb3ff950a7496959eec3b53e0e34f13d01662c80/test/e2e_node/e2e_service.go#L58)):*
<details><summary><i>Previously, vishh (Vish Kannan) wrote…</i></summary>
> I don't get this. Why can't the test binary start processes in a separate process and then run the tests itself?
</details>
@vishh Isn't that what this is doing?

---

*[test/e2e_node/e2e_service.go, line 81 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-mIOLzD7lOee13bHr:-KP-mIOMa9RAmzV4h9bg:b-mtngac) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_service.go#L81)):*
> ```Go
> 		"--cgroups-per-qos="+strconv.FormatBool(framework.TestContext.CgroupsPerQOS),
> 		"--manifest-path", manifestPath,
> 		"--eviction-hard", framework.TestContext.EvictionHard,
> ```

Is it the case that you just want to forward most of the flags? If so, consider just using `pflag.VisitAll(...)` to copy all the flags.

---

*[test/e2e_node/e2e_service.go, line 147 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-nJN2-HevZUl752dh:-KP-nJN2-HevZUl752di:b-ry5fdn) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_service.go#L147)):*
> ```Go
> 	}
> 	return nil
> }
> ```

Isn't this code copied from somewhere? Can you share this in both places, or extract a common method?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 59 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-j3LSF6mR67R7FdKe:-KP-rLeQjVQlrx5J2cbB:b-1xxuq0) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L59)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> nit: (grammar) s/avoid/prevent/
</details>
Thanks~ Correct my grammar anytime. I really need that. :)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 53 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-kEFz8U-Ik7NB6Tof:-KP-r1rZ6MZ9x7d0zQyy:br562s3) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L53)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> The relationship of `start-services` and `start-services-only` is a little confusing. Would it make more sense to have a `run tests` flag that could be used with `start-services`?
</details>
Hm...Yeah...It is confusing.
`start-services` tells the test whether it should start services in before suite.
`start-services-only` tells the test binary to only run services this time.

When `start-services` is specified, the test will start another process in`start-services-only` mode to only start services.
I'll change the name of these flags.

And in fact I'm not sure whether we need `start-services` and `stop-services` flags at all, we never use them.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e/framework/test_context.go, line 72 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_y8-r3-72:-KP0-Zp4T5JN-DjkXysg:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e/framework/test_context.go#L72)):*
<details><summary><i>Previously, ronnielai wrote…</i></summary>
> Could you put in the comment about when DisableKubenet will be removed?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e/framework/test_context.go, line 160 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_y8-r3-160:-KP02TcIQ4Yt14HsZvex:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e/framework/test_context.go#L160)):*
<details><summary><i>Previously, ronnielai wrote…</i></summary>
> Could you elaborate on "set outside"? Outside of what?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e/framework/test_context.go, line 161 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-iTDaJNBFpc4fXdFB:-KP02Vy5gVjek2OCcvst:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e/framework/test_context.go#L161)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Isn't `(default false)` added automatically? If you use `--help` is it shown twice?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 53 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-kEFz8U-Ik7NB6Tof:-KP02XxBlwE7uaUf2uLi:bo5tbwu) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L53)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary>
> Hm...Yeah...It is confusing.
> `start-services` tells the test whether it should start services in before suite.
> `start-services-only` tells the test binary to only run services this time.
> 
> When `start-services` is specified, the test will start another process in`start-services-only` mode to only start services.
> I'll change the name of these flags.
> 
> And in fact I'm not sure whether we need `start-services` and `stop-services` flags at all, we never use them.
</details>
I've changed the original flag to `runServicesMode` and point out that `If true, run all services (etcd, apiserver) in current process, and not run test.`
Hope that will make things a bit clear. :)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 59 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-j3LSF6mR67R7FdKe:-KP03sgwOymcs9ngfrv1:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L59)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary>
> Thanks~ Correct my grammar anytime. I really need that. :)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 61 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-jTlwxx_EJAqaiUKq:-KP03xFq5G7bAWVlp-ZH:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L61)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Flags should not be parsed in init. The problem is, flags could be added at package init time in other packages, and the ordering is hard to track.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 63 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_y9-r1-63:-KP03yfQ7zhge7GfcosG:bcvacs2) ([raw file](https://github.com/kubernetes/kubernetes/blob/bb3ff950a7496959eec3b53e0e34f13d01662c80/test/e2e_node/e2e_node_suite_test.go#L63)):*
<details><summary><i>Previously, vishh (Vish Kannan) wrote…</i></summary>
> Why is this mode needed?
</details>
The original name is misleading. I've changed the name to `runServicesMode`, hope it will make things a bit clear.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 124 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-lFjbX35QKT936Rf8:-KP-s4Z0g6N2w_AxGMUQ:beabvac) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L124)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Can you call RunE2EServices here, or does this do something different? It's confusing...
</details>
Sorry for the confusing.

`startSerives` tells the test suite to start services before test.

The way the test suite starts the services is that it fork another process in the `runServicesMode`, and run services in that process.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

ronnielai commented on test/e2e_node/e2e_node_suite_test.go:
```diff
@@ -124,45 +120,47 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 	if *startServices {
-		e2es = newE2eService(framework.TestContext.NodeName, framework.TestContext.CgroupsPerQOS, framework.TestContext.EvictionHard, shared)
-		if err := e2es.start(); err != nil {
+		e2es = NewE2eServices()
```
Is there a need to have both NewE2eServices() and  StartE2EServices()?

ronnielai replied:
I didn't make myself clear in the previous comment. Since NewE2eServices only returns an empty object, maybe StartE2EServices() can return new the object directly? 

If the new-then-start pattern is preferred, I think the naming should be:
func NewE2eServices() *E2EServices
func (e *E2EServices) Start() (string, error)

Random-Liu commented:
*[test/e2e_node/e2e_service.go, line 81 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-mIOLzD7lOee13bHr:-KP0CDJ3EwT74SimJ-RP:b6llm9) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_service.go#L81)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Is it the case that you just want to forward most of the flags? If so, consider just using `pflag.VisitAll(...)` to copy all the flags.
</details>
If we can move kubelet starting logic out in the future, the only thing we need is `--server-start-timeout` and `--report-dir`. :)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 125 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_y9-r3-125:-KP0Ch_B6h-leZFca4fU:b-3xpkf0) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L125)):*
<details><summary><i>Previously, ronnielai wrote…</i></summary>
> Is there a need to have both NewE2eServices() and  StartE2EServices()? 
</details>
`NewE2eServices` creates the object. `StartE2EServices` is the method of the object. :)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 125 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_y9-r3-125:-KP0KIUAzADUCc2SPp2i:bo5bhig) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L125)):*
<details><summary><i>Previously, ronnielai wrote…</i></summary>
> I didn't make myself clear in the previous comment. Since NewE2eServices only returns an empty object, maybe StartE2EServices() can return new the object directly? 
> 
> 
> 
> If the new-then-start pattern is preferred, I think the naming should be:
> 
> func NewE2eServices() *E2EServices
> 
> func (e *E2EServices) Start() (string, error)
</details>
This totally makes sense... I should change it to `Start` and `Stop`, bad at naming. :(
Thanks!

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 125 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_y9-r3-125:-KP0Kl_pqv_qT6t33dWM:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L125)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary>
> This totally makes sense... I should change it to `Start` and `Stop`, bad at naming. :(
> Thanks!
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_service.go, line 147 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-nJN2-HevZUl752dh:-KP0CTnIz1BSsgY_3xKE:bbeh3r0) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_service.go#L147)):*
<details><summary><i>Previously, timstclair (Tim St. Clair) wrote…</i></summary>
> Isn't this code copied from somewhere? Can you share this in both places, or extract a common method?
</details>
My plan is to remove the original function later after I statically link everything. :)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 103 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_y9-r3-103:-KP04NrfCc3pJLqSNJYb:b-8qbjxf) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_node_suite_test.go#L103)):*
<details><summary><i>Previously, ronnielai wrote…</i></summary>
> Could you make it a method since it's called twice?
</details>
Changed to only initialize once.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_node_suite_test.go, line 147 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KOgWpvkCZFLvWDba_y9-r1-147:-KPA_N1CCMAGue5HoEy2:b-8jo8na) ([raw file](https://github.com/kubernetes/kubernetes/blob/bb3ff950a7496959eec3b53e0e34f13d01662c80/test/e2e_node/e2e_node_suite_test.go#L147)):*
> change
Changed to only initialize it once.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_service.go, line 147 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KP-nJN2-HevZUl752dh:-KPA_ZfBd71c0_zxZK6i:bnexv09) ([raw file](https://github.com/kubernetes/kubernetes/blob/7c920a22e38bd78afc685139126f5c3521c084f4/test/e2e_node/e2e_service.go#L147)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary>
> My plan is to remove the original function later after I statically link everything. :)
</details>
Changed to let it share the same code to start/stop process with specific command.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
*[test/e2e_node/e2e_service.go, line 414 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116#-KPAdpTaoXf6E7fPIkJG:-KPAdpTaoXf6E7fPIkJH:bja30pb) ([raw file](https://github.com/kubernetes/kubernetes/blob/5a98fbe88ea40f353aaed3d45c923fc475758c65/test/e2e_node/e2e_service.go#L414)):*
> ```Go
> 
> // server manages a server started and killed with commands.
> type server struct {
> ```

I've cleaned up the original `healthCheckCommand` and `killCmd` into `server`. Currently in this PR all servers (apiserver, etcd, kubelet) are started with this. And the "separate process" launching all e2e services is also started with this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30116)*
<!-- Sent from Reviewable.io -->

Random-Liu commented:
@timstclair @ronnielai I've addressed the comments. Because one of the comment https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30116#-KP-nJN2-HevZUl752dh, I refactored the code a bit to add a object `server` to unify the functionality for `healthCheckCommand` and `killCommand`, and start all servers (apiserver, etcd, kubelet) and the "separate process" with the same code.

ronnielai commented on test/e2e_node/e2e_service.go:
```diff
@@ -33,22 +34,122 @@ import (
 
+// E2EServices starts and stops e2e services in a separate process. The test uses it to start and
+// stop all e2e services.
+type E2EServices struct {
```
The logic of E2EServices should be in a separate file, shouldn't it?

Random-Liu replied:
Yeah, will do that in a following code refactoring PR. https://github.com/kubernetes/kubernetes/pull/30540

ronnielai commented on test/e2e_node/e2e_service.go:
```diff
@@ -33,22 +34,122 @@ import (
+// RunE2EServices actually start the e2e services. This function is used to
+// start e2e services in current process. This is only used in run-services-mode.
+func RunE2EServices() {
+	e := newE2eService()
```
There are the usage of E2E and E2e. I think they should be consistent.

Random-Liu replied:
The name was there before, but yeah, we should use E2E, will change it. Thanks :)

timstclair commented:
LGTM to unblock you. We can iterate in later PRs if needed.

Random-Liu commented:
@timstclair Thanks a ton!! Will apply labels after addressing @ronnielai's comments.

Random-Liu commented:
Addressed the comments. Self apply LGTM. Thank all of you for reviewing! ;)

Random-Liu commented:
Mark P2 because there are 5 PRs relying on this one. And one of them is marked P1 to resolve the gci flake.

Random-Liu commented:
@k8s-bot test this issue #30462.

Random-Liu commented:
@k8s-bot test this issue #30462.

Random-Liu commented:
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/30719

Random-Liu commented:
@k8s-bot test this issue #30719

Random-Liu commented:
@k8s-bot test this issue #IGNORE.
